server/test/auth/ClientIpResolverTest.java

```diff
@@ -36,41 +43,83 @@ public void resolveClientIp_forwarded() {
+    when(MOCK_SETTINGS_MANIFEST.getClientIpType()).thenReturn(Optional.of("FORWARDED"));
+    var clientIpResolver = new ClientIpResolver(MOCK_SETTINGS_MANIFEST);
 
     var request = new FakeRequestBuilder().withXForwardedFor("3.3.3.3, 2.2.2.2").build();
 
     assertThat(clientIpResolver.resolveClientIp(request)).isEqualTo("2.2.2.2");
   }
 
+  @Test
+  public void resolveClientIp_overload_multipleProxies_singleHeader() {
+    when(MOCK_SETTINGS_MANIFEST.getNumTrustedProxies()).thenReturn(Optional.of(2));
+    when(MOCK_SETTINGS_MANIFEST.getClientIpType()).thenReturn(Optional.of("FORWARDED"));
+    var clientIpResolver = new ClientIpResolver(MOCK_SETTINGS_MANIFEST);
+
+    var request = new FakeRequestBuilder().withXForwardedFor("3.3.3.3, 2.2.2.2, 1.1.1.1").build();
+
+    assertThat(clientIpResolver.resolveClientIp(request)).isEqualTo("2.2.2.2");
+  }
+
+  @Test
+  public void resolveClientIp_overload_multipleProxies_multipleHeaders() {
+    when(MOCK_SETTINGS_MANIFEST.getNumTrustedProxies()).thenReturn(Optional.of(2));
+    when(MOCK_SETTINGS_MANIFEST.getClientIpType()).thenReturn(Optional.of("FORWARDED"));
+    var clientIpResolver = new ClientIpResolver(MOCK_SETTINGS_MANIFEST);
+
+    var request =
+        new FakeRequestBuilder()
+            .withXForwardedFor("3.3.3.3, 2.2.2.2")
+            .withXForwardedFor("1.1.1.1")
+            .build();
+
+    assertThat(clientIpResolver.resolveClientIp(request)).isEqualTo("2.2.2.2");
+  }
+
+  @Test
+  public void resolveClientIp_overload_proxyConfigurationError() {
+    when(MOCK_SETTINGS_MANIFEST.getNumTrustedProxies()).thenReturn(Optional.of(2));
+    when(MOCK_SETTINGS_MANIFEST.getClientIpType()).thenReturn(Optional.of("FORWARDED"));
+    var clientIpResolver = new ClientIpResolver(MOCK_SETTINGS_MANIFEST);
+
+    var request = new FakeRequestBuilder().withXForwardedFor("1.1.1.1").build();
+
+    assertThatThrownBy(() -> clientIpResolver.resolveClientIp(new PlayWebContext(request)))
+        .isInstanceOf(ConfigurationException.class)
+        .hasMessage(
+            "The configured number of trusted proxies (2) is greater than the number of forwarding"
+                + " hops (1)");
+  }
+
   @Test
   public void getClientIpType_forwarded() {
-    var clientIpResolver =
-        new ClientIpResolver(
-            ConfigFactory.parseMap(ImmutableMap.of("client_ip_type", "FORWARDED")));
+    when(MOCK_SETTINGS_MANIFEST.getNumTrustedProxies()).thenReturn(Optional.of(1));
+    when(MOCK_SETTINGS_MANIFEST.getClientIpType()).thenReturn(Optional.of("FORWARDED"));
+    var clientIpResolver = new ClientIpResolver(MOCK_SETTINGS_MANIFEST);
 
     assertThat(clientIpResolver.getClientIpType()).isEqualTo(ClientIpType.FORWARDED);
   }
 
   @Test
   public void getClientIpType_direct() {
-    var clientIpResolver =
-        new ClientIpResolver(ConfigFactory.parseMap(ImmutableMap.of("client_ip_type", "DIRECT")));
+    when(MOCK_SETTINGS_MANIFEST.getNumTrustedProxies()).thenReturn(Optional.of(1));
```

Comment:
*question* The purpose here is to test that the NumTrustedProxies value is ignored for `DIRECT`, right?

Reply:
I was really just trying to replicate the existing test using `SettingsManifest` rather than `Config`.

Reply:
But the current test doesn't set that setting value at all?

(I don't have strong opinions here so feel free to resolve, I was just curious.)